File: Misc samples/CPP/sample_id.cpp
/*! \copyright Sertainty Corporation, 2018.  All Rights Reserved.

    \file     sample_id.cpp

    \brief  Sertainty sample program using native C++ language interface.  Uses XML to create ID.

    \author   Greg Smith
    \date     06/20/2018

    \note     Application expects to find the necessary source files in the current working folder.
 */

#include "uxpfile.h"
#include "uxplicense.h"
#include "uxpcredential.h"
#include "uxlbytearray.h"
#include "uxpid.h"
#include "uxpidmanager.h"

#include <iostream>

/**
 * Response from the user
 * @param ch Challenge object
 * @param prefix Response prefix
 * @return True if accepted.  False if canceled
 */
static int getResponse(uxp::challenge &ch, const std::string &prefix);

/**
 * @brief Main entry point
 * @param argc Number of arguments
 * @param argv Array of argument tokens
 * @return 0 - success, 1 - error
 */
int main(int argc, char *argv[])
{
  bool licensed;
  std::string outfile("$(SAMPLE)/temp/sample.uxp");
  uxl::bytearray idbuf, outbuf, buffer;
  uxp::openVFHandle *handle;
  int status;
  bool done;

  std::cout << std::endl << std::endl << "Sample C++ UXP Application using ID" << std::endl << std::endl;

  /* Set up log file and initialize library.  Must do this before any active calls. */

  try
  {
    uxp::sys::setLogFile("SampleCPP", "Sample C++ 2.0.0");

    /* Attempt to find the license file. */

    try
    {
      uxp::sys::initializeLibrary(argc, argv,
                                  "sertainty.lic",
                                  "SertintyONE");
      licensed = true;
    }
    catch (uxl::exception &)
    {
      licensed = false;
    }

    uxp::sys::setLogging(uxpLogInfo);

    if (!licensed)
    {
      std::cout << "UXP-432, A valid license was not found." << std::endl;

      return 1;
    }

    uxp::file data;

    /* Create a new ID from the sampleid.xml ID definition */

    uxp::sys::fileReadAll("$(SAMPLE)/SampleData/sampleid.xml", idbuf);

    uxp::idManager::publishToFile("$(SAMPLE)/temp/sampleid.iic", idbuf, ModifierReplace);

    /* Create new UXP from ID */

    std::cout << "Opening new UXP: " << outfile << std::endl;

    data.openNewFile(outfile, "$(SAMPLE)/temp/sampleid.iic", IdFile, ModifierReplace, 0);
    data.addVirtualFileFromFile("data.pdf", std::string("$(SAMPLE)/SampleData/data.pdf"), -1, -1, 0);

    /* Read the virtual file and write it back out to disk */

    std::cout << "Extracting data.pdf to copy1.pdf" << std::endl;

    handle = data.openVirtualFile("data.pdf", ModeRead);

    buffer.assign("$(SAMPLE)/temp/copy1.pdf");
    uxp::sys::processStandardMacros(buffer);

    FILE *f1 = fopen(buffer.data(), "wb");

    while (data.readVirtualFile(handle, outbuf))
    {
      fwrite(outbuf.data(), static_cast<size_t>(outbuf.size()), 1, f1);
    }

    fclose(f1);
    data.closeVirtualFile(handle);

    if (data.compareExternalFile("data.pdf", "$(SAMPLE)/temp/copy1.pdf"))
      std::cout << "Comparison of data.pdf to copy1.pdf: successful" << std::endl;
    else
      std::cout << "Comparison of data.pdf to copy1.pdf: failed" << std::endl;

    /* Close the UXP */

    data.close();

    /* Reopen the UXP ... includes authentication. */

    std::cout << "Opening new UXP" << std::endl << std::endl;

    std::cout << "Credentials necessary to access UXP:" << std::endl << std::endl;
    std::cout << "  Username = SampleUser@myemail.com" << std::endl;
    std::cout << "  Challenge 1 = Response 1" << std::endl;
    std::cout << "  Challenge 2 = Response 2" << std::endl;
    std::cout << "  ... " << std::endl;
    std::cout << "  Challenge 10 = Response 10" << std::endl << std::endl;

    data.openFile(outfile, ShareAll);

    done = false;

    while (!done)
    {
      status = data.authenticate();

      switch (status)
      {
        case StatusAuthorized:
          std::cout << "You are authorized" << std::endl;
          done = true;
          break;

        case StatusNotAuthorized:
          std::cout << "You are not authorized" << std::endl;
          return 1;

        case StatusChallenged:
          for (auto &t_ch: data.getChallenges())
          {
            uxp::challenge ch(t_ch);

            status = getResponse(ch, "Response");
            if (!status)
            {
              std::cout << "Canceled" << std::endl;
              data.close();

              return 1;
            }

            data.addResponse(ch);
          }
          break;

        default:
          break;
      }
    }

    /* Read the virtual file and write it back out to disk */

    std::cout << std::endl << "Extracting data.pdf to copy2.pdf" << std::endl;

    handle = data.openVirtualFile("data.pdf", ModeRead);

    buffer.assign("$(SAMPLE)/temp/copy2.pdf");
    uxp::sys::processStandardMacros(buffer);

    f1 = fopen(buffer.data(), "wb");

    while (data.readVirtualFile(handle, outbuf))
    {
      fwrite(outbuf.data(), static_cast<size_t>(outbuf.size()), 1, f1);
    }

    fclose(f1);
    data.closeVirtualFile(handle);

    if (data.compareExternalFile("data.pdf", "$(SAMPLE)/temp/copy2.pdf"))
      std::cout << "Comparison of data.pdf to copy2.pdf: successful" << std::endl;
    else
      std::cout << "Comparison of data.pdf to copy2.pdf: failed" << std::endl;

    data.close();

    std::cout << std::endl << "Sample finished running" << std::endl;
  }
  catch (uxl::exception &e2)
  {
    std::cout << "Error: " << e2.getErrorText() << std::endl;
  }

  return 0;
}

/**
 * Response from the user
 * @param ch Challenge object
 * @param prefix Response prefix
 * @return True if accepted.  False if canceled
 */
static int getResponse(uxp::challenge &ch, const std::string &prefix)
{
  std::string value;
  uxl::string prompt;

  prompt = ch.getPrompt();

  ch.startTimer();

  if (prompt == "Username")
  {
    value = "SampleUser@myemail.com";
  }
  else
  {
    value = prefix + " " + prompt.mid(10);
  }

  ch.endTimer();
  ch.setValue(value);

  return 1;
}

File: Misc samples/CS/sample_id.cs
using System;
using Sertainty;

namespace Sertainty
{
  using System.IO;
  
  class SampleID
  {
    static void Main (string[] args)
    {
      Console.WriteLine("Sample C# Data Application using ID");
      
      string idXmlSpec = "sampleid.xml";
      string idFileSpec = "sampleid.iic";
      string uxpFileSpec = "sample.uxp";
      string dataPdfSpec = "data.pdf";
      string dataPdf2Spec = "data2.pdf";
      string copy1Spec = "copy1.pdf";
      string copy2Spec = "copy2.pdf";
      
      UxpFile appHandle;
      
      ByteArray buffer = new ByteArray();
      Sys.CallStatus callstatus = new Sys.CallStatus();
      
      Sys.SetLogFile("Sample C#", "Sample C# 1.0");
      
      long ret = Sys.InitializeLibrary(buffer, args.LongLength, args, "sertainty.lic", "SertintyONE", "Sample C#", "Sample C# 1.0");
      if (ret == 0)
      {
        Console.WriteLine("Error initializing environment: {0}", buffer);
      }
      else
      {
        Console.WriteLine("Library Initialized");
        Sys.FileReadAll(callstatus, idXmlSpec, buffer);
        
        if (callstatus.HasError)
        {
          Console.WriteLine("Error reading ID: {0}", callstatus.ErrorMessage);
        }
        else
        {
          Console.WriteLine("{0} read", idXmlSpec);
          string doc = buffer.ToString();
          Id.PublishIdToFile(callstatus, idFileSpec, doc, (long)Modifiers.Replace);
          
          if (callstatus.HasError)
          {
            Console.WriteLine("Error creating ID: {0}", callstatus.ErrorMessage);
          }
          else
          {
            Console.WriteLine("{0} created", idFileSpec);
            appHandle = new UxpFile();
            appHandle.OpenNewFile(uxpFileSpec, idFileSpec, 3, (long)Modifiers.Replace, 0);
            
            if (appHandle.HasError)
            {
              Console.WriteLine("Error creating Data: {0}", appHandle.ErrorMessage);
            }
            else
            {
              Console.WriteLine("{0} created", uxpFileSpec);
              appHandle.AddVirtualFile("data.pdf", dataPdfSpec, -1, -1, (long)Modifiers.Compress);
              
              if (appHandle.HasError)
              {
                Console.WriteLine("Error creating virtual file: {0}", appHandle.ErrorMessage);
              }
              else
              {
                Console.WriteLine("{0} added", dataPdfSpec);
                appHandle.AddVirtualFile("data2.pdf", dataPdf2Spec, -1, -1, (long)Modifiers.Compress);
                
                if (appHandle.HasError)
                {
                  Console.WriteLine("Error creating virtual file: {0}", appHandle.ErrorMessage);
                }
                else
                {
                  Console.WriteLine("{0} added", dataPdf2Spec);
                  
                  //Now open the first virtual file and write it out to a temporary file.
                  VirtualFile fileHandle = appHandle.OpenVirtualFile("data.pdf", Mode.ReadOnly);
                  
                  if (appHandle.HasError)
                  {
                    Console.WriteLine("Error opening virtual file: {0}", appHandle.ErrorMessage);
                  }
                  else
                  {
                    Console.WriteLine("{0} opened", "data.pdf");
                    
                    Console.WriteLine("Reading data.pdf in loop ...");
                    FileStream sw = new FileStream(copy1Spec, FileMode.Create);
                    
                    while (fileHandle.Read(buffer, 1000) > 0)
                    {
                      int len = (int)buffer.Size;
                      byte[] data = buffer.GetBytes(len);
                      //long len = uxpba_getSize(buffer);
                      sw.Write(data, 0, len);
                    }
                    sw.Close();
                    fileHandle.Close();
                    Console.WriteLine("Finished reading data.pdf");
                    
                    if (appHandle.CompareExternalFle("data.pdf", copy1Spec))
                    {
                      Console.WriteLine("Comparison of data.pdf to copy1.pdf: successful");
                    }
                    else
                    {
                      Console.WriteLine("Comparison of data.pdf to copy1.pdf: failed");
                    }
                    
                    //Close the UXP. This will delete the handle as well
                    
                    appHandle.Close();
                    
                    //Reopen the Data ... includes authentication
                    
                    Console.WriteLine("Opening new Data");
                    
                    appHandle.Open(uxpFileSpec, UxpMode.ShareAll);
                    
                    if (appHandle.HasError)
                    {
                      Console.WriteLine("Error opening UXP: {0}", appHandle.ErrorMessage);
                    }
                    else
                    {
                      Console.WriteLine("Credentials necessary to access UXP:");
                      Console.WriteLine("  Username = SampleUser@myemail.com");
                      Console.WriteLine("  Challenge 1 = Response 1");
                      Console.WriteLine("  Challenge 2 = Response 2");
                      Console.WriteLine("  ... ");
                      Console.WriteLine("  Challenge 10 = Response 10\n");
                      
                      bool done = false;
                      bool authorized = false;
                      AuthorizationStatus status;
                      
                      while (!done)
                      {
                        status = appHandle.Authenticate(false);
                        
                        switch (status)
                        {
                          case AuthorizationStatus.Authorized:
                            Console.WriteLine("You're authorized");
                            done = true;
                            authorized = true;
                            break;
                            
                          case AuthorizationStatus.NotAuthorized:
                            Console.WriteLine("You're not authorized");
                            authorized = false;
                            done = true;
                            break;
                            
                          case AuthorizationStatus.Challenged:
                            for (int i = 0; i < appHandle.ChallengeCount; i++)
                            {
                              Challenge ch = appHandle.GetChallenge(i);
                              getResponse(ch);
                              appHandle.AddResponse(ch);
                            }
                            break;
                          
                          default:
                            break;
                        }
                      }
                      
                      if (authorized)
                      {
                        Console.WriteLine("Extracting data.pdf to copy2.pdf");
                        
                        fileHandle = appHandle.OpenVirtualFile("data.pdf", Mode.ReadOnly);
                        
                        if (appHandle.HasError)
                        {
                          Console.WriteLine("Error opening virtual file: {0}",  appHandle.ErrorMessage);
                        }
                        else
                        {
                          sw = new FileStream(copy2Spec, FileMode.Create);
                          while (fileHandle.Read(buffer, 1000) > 0)
                          {
                            int len = (int)buffer.Size;
                            byte[] data = buffer.GetBytes(len);
                            sw.Write(data, 0, len);
                          }
                          sw.Close();
                          fileHandle.Close();
                        }
                      }
                    }
                  }
                }
              }
            }
            
            appHandle.Close();
          }
          
          Console.WriteLine("Sample finished running. Press any key to exit...");
          
          Console.ReadKey();
        }
      }
    }
    
    private static void getResponse(IntPtr ch_handle)
    {
      ByteArray byteArray;
      Challenge ch;
      ch = Challenge.fromHandle(ch_handle);

      byteArray = ch.GetPrompt();
      ch.StartTimer();

      if (byteArray.ToString() == "Username")
      {
        ch.SetValue("SampleUser@myemail.com");
      }
      else
      {
        string challenge = byteArray.ToString();
        string[] list = challenge.Split(' ');
        string response = "Response " + list[1].ToString();
        ch.SetValue(response);
      }

      ch.EndTimer();
    }
  }
}
